### DISCUSSION TITLE
i feel hard to use proxy.py to write a real proxy for crossing government wall

### DISCUSSION THREAD

**Discussion by sjj151461** [abhinavsingh/proxy.py#1499] [General]
that's the real value, but proxy.py is hard to use, designed  with confusion

can anybody show me how to write a encrypt url proxy, it has two nodes, nodeA just forward encrypted url(for crossing government wall) and tcp steam to nodeB, nodeB decrypt url and connect to website, then put everything back to nodeA

i acctually write one with proxy.py, most time it's ok, but it's buggy some time
how to test：setup nodeA in china，setup nodeB in usa or somewhere， use nodeA for your chrome proxy，visit www.google.com or www.youtube.com

**Comment by abhinavsingh** [comment]
@sjj151461 Can you expand on "it's buggy some time".   What exactly is buggy?

`proxy.py` purpose is not to bypass any gov firewall, but yes, generally `proxy.py` can be used for such purposes too.  However, all depends upon what kinds of restrictions need to be bypassed.   You can try following approaches:

1) Run `proxy.py` locally with DoH enabled.    This might work as DNS resolution is done over HTTP2.
2) Run `proxy.py` in US region, run another `proxy.py` locally that points `--proxy-pool` to the one running in US region.   Point your system to use local proxy server.   This solution will internally channel all requests to remove proxy.
3) Run `proxy.py` in US region and point your system to this proxy.   This solution will directly send all traffic from your system to remote proxy.

**Comment by sjj151461** [comment]
i want to talk about library API design, that's the main problem when i try to use proxy.py to solve my problem.
good design block useless info for user, you only need know API，then the job is done.
the more user need to dig into your system, the bad the API design is. So the API designer need to know user's demand, and which tools need to be prepared, and whatever else is hidden, and what is hidden is designer work's real value. For example, start a car you only need to press a engine start button，anything else is hidden, car engineers does all the work for you.

after try about 4 hours learning how to use proxy.py, and digging into such spider web like code, i still can not startup the car. then i use raw python3 write my own proxy, after 8 hours, my job is mostly done, i feel python3 offer very good API，and i only need to know very little information to use them, and i never need to dig into their code.

so before you design a library, think about what problem you want to solve, and people NOT FAMILIER with your library need how much time to solve their problem. A great library and a library can use is different.

a good proxy library for me, i need it to hide the complex real net environment, and better it is designed with layer，clear data flow，never over encapsulation

**Reply by abhinavsingh** [reply]
Thank you for the feedback.   `proxy.py` as it stands today is often used as proxy server directly w/o having to hack the internals.   But I agree, we can certainly improve upon the documentation from developer point of view.
